src/CacheWithResilience.php

```diff
@@ -0,0 +1,165 @@
+<?php
+declare(strict_types=1);
+
+namespace BBC\ProgrammesCachingLibrary;
+
+use DateTime;
+use DateInterval;
+use DateTimeInterface;
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * An implementation of stale-if-error for the Cache library.
+ * This class is compatible with BBC\ProgrammesCachingLibrary\Cache, the only difference when calling the functions
+ * is that you can call getItem() with an extra parameter to get the stale value.
+ */
+class CacheWithResilience extends Cache
+{
+    /** @var LoggerInterface  */
+    private $logger;
+
+    /** @var \Exception[]  */
+    private $whitelistedExceptions;
+
+    // How long items will persist in the cache server.
+    // This also mean for how long we could serve stale content in case of issues
+    private const RESILIENCE_TTL = 21600; // 6h
+
+    public function __construct(
+        LoggerInterface $logger,
+        CacheItemPoolInterface $cachePool,
+        string $prefix,
+        array $cacheTimes = [],
+        array $whitelistedExceptions = []
+    ) {
+        $this->logger = $logger;
+        $this->whitelistedExceptions = $whitelistedExceptions;
+        parent::__construct($cachePool, $prefix, $cacheTimes);
+    }
+
+    /**
+     * This function get the item if exist in memory and the TTL
+     *
+     * @param string $key
+     * @param bool $returnStaleValue
+     * @return CacheItemInterface
+     */
+    public function getItem(string $key, bool $returnStaleValue = false): CacheItemInterface
+    {
+        $key = $this->standardiseKey($key);
+        if ($this->flushCacheItems) {
+            $this->cachePool->deleteItem($key);
+        }
+
+        $item = $this->cachePool->getItem($key);
+        if ($item->isHit()) {
+            $value = $item->get();
+            // we have 2 TTL's, one of them is stored with the item as $value['ttl'] and is the time
+            // when we want the item to be consider as expired, the second one is the real TTL that we tell redis.
+            // The second one is only mean to be used in case of issues like DB down and only requires to
+            // set $returnStaleValue to true
+            if (false === $returnStaleValue && new DateTime() > $value['ttl']) {
+                return new CacheItem();
+            }
+        }
+
+        // in setItem() value gets wrapped into an array with the ttl. set the item to its original value
+        $val = $item->get();
+        $item->set($val['value']);
+
+        return $item;
+    }
+
+    /**
+     * @param CacheItemInterface $item
+     * @param mixed $value
+     * @param int|string|DateTimeInterface $ttl
+     *   TTL in seconds, or a constant from CacheInterface, or a DateTime to expire at
+     * @return bool
+     */
+    public function setItem(CacheItemInterface $item, $value, $ttl): bool
```

Comment:
I'd suggest decide on either "time to live" (seconds) or "expire at" (future date) but do not try to accommodate both. with dates especially it can get crazy hard very quickly. my vote is for `ttl` in seconds.

Reply:
This function is already defined in `src/Cache.php `, I copied it and only edit the internals to add the resilience mechanims. Seconds and Dates were already possible so I can't change it wothout breaking compatibility.

On the other hand our caching library es PSR-compliant and you can cahnge the driver to something else symfony support and will continue to work